### ISSUE TITLE
Provider suggestion: Dinasty Reader

### ISSUE THREAD

**Issue by XellossMetallium** [Deathspike/mangarack#47]
@Deathspike I don't know if you mean the providers included as a closed number, but if not there is only one I miss, since Batoto, KM and Mangafox cover almost everything.
So.. the main reason is that the series of http://dynasty-scans.com/ are exclusive of this site. Some are nsfw (thus excluded by batoto policies) but there is also a good number of high quality scans (on par with batoto) from scanlators that, for a reason or another, do not consider batoto a viable option. 
Moreover there are a lot of doujinshi properly tagged (actually tagging system is very detailed).

The site provides download links for chapter (limited number per-day) but has no restriction in browsing (so it should be perfectly compatible with Mangarack).
- Series are provided with chapter lists (http://dynasty-scans.com/series/shinozaki_san_ki_wo_ota_shika_ni )
- One-Shots are on a single layer ( http://dynasty-scans.com/chapters/desk_war ), but author page lists them together (http://dynasty-scans.com/authors/dowman_sayman).
- Unique links for reading lists are available and optionally public (http://dynasty-scans.com/lists/115624-test-list), so no log in needed (related to #36 )

My best regards!

**Comment by Deathspike** [comment]
I will maintain Batoto, Mangafox and Kissmanga, but that does not mean I won't allow pull requests with new providers if said person is willing to continue supporting that new provider! The source code dealing with provider registration is at [in component-core/providers](https://github.com/Deathspike/mangarack/blob/master/mangarack-component-core/src/providers/default.ts). All that is required is defined in the [interfaces](https://github.com/Deathspike/mangarack/tree/master/mangarack-component-core/src/typings/providers). Of course, this may all be very intimidating to non-programmers, but it generally boils down to "how the page is structured". MangaRack just looks at the markup of a page and extracts bits and pieces from it.

I'm perfectly happen to answer questions and provide guidance when necessary, but unfortunately time does not allow me to implement new providers at this point. As you can see from the commit history, I haven't even been pushing code for ~2 months now. I will focus my time towards the library features first and foremost, so any new features not directly related to that are put on the back burner for me. But again, I totally welcome other people wanting to work on new providers, and would be happy to help you on your way!

So, up for grabs. 

P.S: This site looks interesting. It has tooons of Doujinshi which is very cool :+1:

**Comment by XellossMetallium** [comment]
Understood!
Intimidating is the right word, but I'll try to figure it out after my exam session, in 2-3 weeks. With some luck I will not need to bother you (too much) :)
Thanks!

**Comment by yanandcoffee** [comment]
Don't mean to hijack your efforts, @XellossMetallium, but are you still working on this? 

I'm interested in picking this up if you are not, as I'm also a fan of stuff from Dynasty Reader. If you have some work on this already and may need some help, I'd also be glad to help out.

**Comment by XellossMetallium** [comment]
No problem at all, @yanandcoffee , thanks for your consideration!
I starded checking out the code during summer but I am still far from understanding it, so I do not even have anything to give as a base. I am afraid to be much below a complete beginner :(
Please pick this up along with my gratitude! I'm looking forward to your work.

**Comment by yanandcoffee** [comment]
👍  Thanks for letting me know, @XellossMetallium! I'll be picking this up then. I haven't dealt with TypeScript before, but really wanted to try it out, so this should be interesting.

Any advice on how to test whether my code works or not, @Deathspike?

**Comment by Deathspike** [comment]
Sure. The most important part of scraping the information. First use your browser and developer tools:

1. Open the chapter listing in your browser and disable JavaScript. You'll have one of the options:
1a. The page still contains the needed information. You're good to go.
1b. The page no longer contains the needed information. You'll have to emulate JS to get the information.
2. See if the chapter listing references jQuery. MangaRack uses Cheerio, which uses the same API. So:
2a. The page contains jQuery. You're good to go.
2b. The page does not contain jQuery. Add it to the page so you'll have access to the jQuery API.
3. Now you can start writing jQuery selectors. Reference MangaFox series.ts for needed selectors.

You should aim to make the selectors as non-specific as possible. For example, it's much better to have `.series_preview` rather than `.some-container > .some-page > div > .series_preview`, because the former is much more likely to survive changes on the provider. At this point you haven't even touched MangaRack, but you should have a nice collection of selectors for all the information you need, including how you would want to traverse pages and get the images there. Now you're ready to start adding this to MangaRack:

1. Clone the master branch and follow the installation instructions. Don't use the development branch!
2. In `mangarack-component-core/src/providers` clone the most similar provider into a new folder.
3. Hook up your new provider in `mangarack-component-core/src/providers/default.ts`.
4. Change the `name` and `isSupported` function of your copy to accommodate your new provider.

MangaRack merely throws an series address into each provider, and the first to respond with 'Yeah I can deal with this' is used to process that particular series. So make sure your provider implementation only accepts valid chapter listing addresses of your provider and nothing else. Now it's a matter of tweaking the following structure in series, chapters and pages to fit your provider (you'll be using your selectors and page traveral knowledge here). 

You'll need a TS compiler in your editor (for example Atom with TypeStrong), but it will protect you from mistakes. Go through the structure of `ISeries`, `IChapter`, `IPage` and `IProvider` to understand how MangaRack expects your provider implementation to behave. The naming should be relatively straight forward. 

Let me know if you have further questions or need help!

**Comment by yanandcoffee** [comment]
@Deathspike Thanks a lot for the information! I had wrote some code already, and I'm glad you've confirmed what I needed to do! I'll definitely ask if I need more help.

**Comment by yanandcoffee** [comment]
Just a quick update that I'm done with adding DynastyReader as a provider and implementing the series downloading as its primary function.

For the one-shots and reading lists support, it may require more discussion around how to approach this here as I know there's been some discussion around filtering, or being able to download a specific chapter or range using mangarack. What do you think of this, @Deathspike?

**Comment by Deathspike** [comment]
I'm not too sure with what you mean with 'one-shots'. Does it also appear under a series listing, or is it something else entirely?

I've thought about the concept of reading lists on various occasions, we've also had a discussion going for it at https://github.com/Deathspike/mangarack/issues/36. The whole idea would be to make this a separate CLI application that merely feeds addresses into mangarack. So, `mangarack-cli < mangarack-list http://example.foo/bar` would be the notation. Or `mangarack-list http://example.foo/bar | mangarack-cli` on Linux (I'm not sure that works on Windows). So, absolutely take a stab at that runnable if you feel up for it!

Discussion about filters is at https://github.com/Deathspike/mangarack/issues/53 which I quite simply have forgotten in the sea of other things I've been working on. I think the specification is quite solid, so do leave your thoughts for it if you see a problem with it for dynasty reader

**Comment by XellossMetallium** [comment]
If I'm not wronk series are under 
/series/title
their chapters are under
/chapters/chapter01
also oneshots are under
/chapters/oneshot
but a "series" level seems to be missing for them.

Lastly both oneshots and series are still listed under
/authors/authorname

Support for the author level would be nice but I suppose needlessy complicated (since it would mean to recognize and deal with both cases); for oneshots it would translate to the ability to download just a single chapter, I think.

**Comment by yanandcoffee** [comment]
@Deathspike Yes, you are correct, the last two does not have a "series" level which means the work I've done so far for provider would not work unless I support the "chapters" level or the "authors" level, which leads into more complexity for the code. Perhaps I will take a stab at it after  the core part of it has been merged in.

Also, I just realized that there is another level, the "anthologies" level which is the same structure as the "series" level with slight differences, but it is under a different URL... `http://dynasty-scans.com/anthologies/<title>`. I'll probably add support to this later as well.
